usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
     return CKR_OK;
 }
 
+/*******************************************************************************
+ *
+ *                    Begin EP11 protected key option
+ */
+
+typedef struct {
+    CK_BBOOL wrap_was_successful;
+    CK_VOID_PTR secure_key;
+    CK_ULONG secure_key_len;
+    CK_BYTE *pkey_buf;
+    size_t *pkey_buflen_p;
+} pkey_wrap_handler_data_t;
+
+/* A wrapped key can have max 132 bytes for an EC-p521 key */
+#define MAX_WRAPPED_KEY_SIZE    (2 * (521 / 8 + 1))
+typedef struct {
+    uint16_t version;
+    uint8_t res0[16];
+    uint32_t cpacf_key_type;
+    uint32_t bit_length;
+    uint64_t token_size;
+    uint8_t res1[8];
+    uint8_t wrapped_key[MAX_WRAPPED_KEY_SIZE];
+    uint8_t res2[50];
+} __attribute__((packed)) wrapped_key_t;
+
+/**
+ * Callback function used by handle_all_ep11_cards() for creating a protected
+ * key via the given APQN (adaper,domain).
+ */
+static CK_RV ep11tok_pkey_wrap_handler(uint_32 adapter, uint_32 domain,
+                                       void *handler_data)
+{
+    CK_BYTE iv[16] = {
+        0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+        0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10
+    };
+    CK_MECHANISM mech = { CKM_IBM_CPACF_WRAP, &iv, sizeof(iv) };
+    pkey_wrap_handler_data_t *data = (pkey_wrap_handler_data_t *) handler_data;
+    target_t target = 0;
+    CK_RV ret;
+
+    if (data->wrap_was_successful)
+        goto done;
+
+    ret = get_ep11_target_for_apqn(adapter, domain, &target, XCP_MFL_PROBE);
+    if (ret != CKR_OK)
+        goto done;
+
+    /* Create the protected key via CKM_IBM_CPACF_WRAP */
+    ret = dll_m_WrapKey(data->secure_key, data->secure_key_len,
+                        NULL, 0, NULL, 0, &mech,
+                        data->pkey_buf, data->pkey_buflen_p,
+                        target | XCP_TGTFL_SET_SCMD);
```

Comment:
Is this really right? You or XCP_TGTFL_SET_SCMD to the target?  Dependent on the Ep11 library version, target is either a numeric value (handle), or a pointer (casted to uint64_t). I guess since protected keys only work with a host lib of >= 3 it should always be a handle, but I still don't think that this could be or'ed to the target handle. Shouldn't this go to the flags passed to get_ep11_target_for_apqn (aside with XCP_MFL_PROBE)?

Reply:
Yes, the target needs the XCP_TGTFL_SET_SCMD flag here. The ep11lib returns an error when leaving this out. The flag is defined in ep11.h as "protected key special command".

Reply:
I am still not fully convinced that this is correct.

target (to which the flag is or'ed) is a target identifyier built by get_ep11_target_for_apqn(). Dependent on if you run with an new or an old EP11 library (i.e. old or new targethandling), this is either a numeric value (where it might be OK to or the flat to), or an address casted to target_t. I doubt that you simply should or a flag into an address.....This will most likely seggfault.

So looks like this support only works with the new target handling, i.w. new EP11 library version.  Is there a check that ensures that this code only is executed when running with the new target handling?

Reply:
Hm yes ... the CPACF_WRAP mechanism works only with the currently latest ep11 lib 3.0.1. At token init, the protected key of a test key is created to save the firmware master key vp. This is only done if the CPACF_WRAP mechanism is supported (checked via ep11tok_is_mechanism_supported). So we never reach routine ep11tok_pkey_skey2pkey and in turn ep11tok_pkey_wrap_handler when running with an old library.

Reply:
Hm, well, whether the CPACF_WRAP mech is supported or not depends on the hardware, not so much on the EP11 library level. So it can happen that you run with new hardware (that supports CPACF_WRAP), but using an old EP11 library. 
However, ep11tok_is_mechanism_supported() seems to check the library version, so this should prevent it from being used with an older library. So this looks OK.
But maybe add a comment to this function about requiring a EP11 library >= V3.

Reply:
Done.